### PR TITLE
Refactor Slingshot to use Typescript

### PR DESCRIPTION
**Pull request by simianhacker**
This PR introduces Typescript and ESLint into the project. The ESLint configuration extends Kibana's configuration so that the developer experience will be similar. Through the course of migrating to Typescript I took this opportunity to add several new features and remove a few things that didn't make sense. 

#### CLI Documentation
I finished building out the CLI interface and CLI documentation for Slingshot. Instead of creating your own scripts with the configuration, you can now create a JSON config file, load it, and override some basic settings via the CLI.

```
❯ bin/slingshot.ts load --help
slingshot.ts load

Loads data into Elasticsearch

Options:
  --version        Show version number                                 [boolean]
  --help           Show help                                           [boolean]
  --config         The path to a config file                 [string] [required]
  --start          Overrides the start date math                        [string]
  --end            Overrides the end date math                          [string]
  --interval       Overrides the event interval                         [number]
  --elasticsearch  Overrides the Elasticsearch connection               [string]
  --auth           Overrides the credentials for Elasticsearch          [string]
  --purge          Deletes data stream before history                  [boolean]
```

#### Removal of Continuous Mode
Slingshot supported a continuous mode along with a history mode. This release removes the continuous mode in favor of setting the history to a date range that spans from the past to the future. For example, users should set a start date expression like `now-7d/d`  and the end date to `now+1d/d`. Events will still appear as if they are being index in real time since Kibana usually includes a date range for every query with a end date set to `Now`.

#### Command Line Overrides
The original prototype interface would allow users to override any of the configuration options. When I converted this project to Typescript I dropped the support for this feature in favor of a limited set of options (`elasticsearch, timerange, purge, auth`). This was done to simplify type checking and give us control over what comes from the user vs. config.

#### No more spike scripts
Prior to this PR, you had to create a seperate script for adding spikes or anomalies. The problem with this approach is that the dataset stops looking like a "REAL" data set. For example, when Metricbeat adds events to the index, they are indexed on a specific interval and the number events per minute is consistent from one minute to the next. Using a second script to insert spike events will add more events then what we would see in Metricbeat and both the normal and spike values will be aggregated. With the new spike definitions, it behaves like Metricbeat and continues to create a consistent number of event per minute so values go from a normal distribution to a spiked distribution without duplicate timestamp entries for the dataset.

#### Included Changes

- Refactor codebase to use Typescript
- Add ESlint support (from Kibana)
- Setup CLI documentation and parsing
- Change templates to be array of events
- Remove continuous mode
- Add purge feature to delete the underlying data stream before indexing
- Add ability to configure multiple data types in one file
- Add spike feature

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -5,21 +5,69 @@
...
   "license": "MIT",
   "bin": {
-    "slingshot": "./src/bin/slingshot.js"
+    "slingshot": "./bin/slingshot.ts"
+  },
+  "scripts": {
+    "typecheck": "tsc --noEmit -p ."
   },
   "dependencies": {
+    "@elastic/datemath": "link:../kibana/packages/elastic-datemath",
```
Comment (jasonrhodes) [inline package.json]: Do we need to update this ref?

**Review by jasonrhodes — CHANGES_REQUESTED** [review]
Overall looks great thank you so much!!! I'm going to run it next week and then give a final thumbs up. Just had one package.json dependency reference change/question.

File: package.json (hunk 1 above)
Comment (simianhacker) [reply package.json]: I changed that last week, forgot to push :D

**Review by jasonrhodes — APPROVED** [review]
LGTM!

**Pull request merged**